### PR TITLE
feat(frontend): pin tokens with balance at top

### PR DESCRIPTION
**Pull request by AntonioVentilii-DFINITY**
Revisiting PR #1779 (that was rolled back with PR #1938)

# Motivation

As final requirement for sorting, we pin the tokens with balance on top, per descending USD amount and after per token amount (if USD value is not provided).

# Changes

- Create sorting function to pin tokens with balance on top.
- Create global derived store `sortedNetworkTokensUi` that pins tokens with balance on top (and sorts them).
- Adapt usage in `TokensSignedIn`.

# Tests

Created test for function `pinTokensWithBalanceAtTop`.

### CODE REVIEW

**Comment by AntonioVentilii-DFINITY** [comment]
@peterpeterparker  I think you will like the last iteration in this. I removed a loop

**Comment by AntonioVentilii-DFINITY** [comment]
> I did not review the sorting algorithm, assuming it's the same as the one we reverted.


Correct, the algorithm is the same, the thing that changed is how we pin the sorted with balance on top. I was looping twice before: once to select the tokens to pin and once to pin them. Now we already select them and put them on top directly during the selection.

**Comment by peterpeterparker** [comment]
What also changed is the fact that we do not use `sortedNetworkTokensUi` to derive any other stores right? That was yesterday root cause of the loop.

File: src/frontend/src/lib/derived/network-tokens.derived.ts [hunk 1]
```diff
@@ -51,3 +52,15 @@ export const enabledNetworkTokensUi: Readable<TokenUi[]> = derived(
 			};
 		})
 );
+
+/**
+ * All tokens matching the selected network or Chain Fusion, with the ones with non-null balance at the top of the list.
+ */
+export const sortedNetworkTokensUi: Readable<TokenUi[]> = derived(
```
[the remark this answers is not in the document]
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/derived/network-tokens.derived.ts:57]: Maybe using the `complexity` flag?

```typescript
/**
 * All tokens matching the selected network or Chain Fusion, with the ones with non-null balance at the top of the list.
 * @complexity High
 * @description This store combines data from multiple sources: network tokens, balances, and exchanges.
 */
```

[the remark this answers is not in the document]
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/derived/network-tokens.derived.ts:57]: Or maybe

` * @dependencies enabledNetworkTokens, balancesStore, exchanges`

**Comment by AntonioVentilii-DFINITY** [comment]
Exactly!

File: src/frontend/src/lib/derived/network-tokens.derived.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (peterpeterparker) [reply src/frontend/src/lib/derived/network-tokens.derived.ts:57]: Maybe or a naming convention that differentiate simple derived stores of composed. Or just `payATTENTION_WARNING_BE_REALLY_SURE_sortedNetworkTokensUi` ;)

[the remark this answers is not in the document]
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/derived/network-tokens.derived.ts:57]: Trying with:

- PR #1984

**Pull request merged**